activesupport/lib/active_support/notifications/fanout.rb

```diff
@@ -91,11 +91,14 @@ def self.new(pattern, listener, monotonic)
           if listener.respond_to?(:start) && listener.respond_to?(:finish)
             subscriber_class = Evented
           else
-            # Doing all this to detect a block like `proc { |x| }` vs
-            # `proc { |*x| }` or `proc { |**x| }`
-            if listener.respond_to?(:parameters)
-              params = listener.parameters
-              if params.length == 1 && params.first.first == :opt
+            # Doing all this to detect a single argument block or callable
+            # like `proc { |x| }` vs `proc { |*x| }`, `proc { |**x| }`,
+            # or `proc { |x, **y| }`
+            if listener.respond_to?(:call)
+              procish = listener.respond_to?(:parameters) ? listener : listener.method(:call)
+              params = procish.parameters
+
+              if params.length == 1 && [:opt, :req].include?(params.first.first)
```

Comment:
I believe `procish.arity == 1 && procish.parameters.length == 1` would be equivalent here. Perhaps I should update to that instead?

`procish.arity` gets us most of the way there, since `proc { |*x| }` has an arity of -1 and `proc { |**x| }` has an arity of 0. The parameter length check would then cover `proc { |x, **y| }` which has an arity of 1 but a parameter length of 2.

Reply:
👍

```ruby
irb(main):001:0> proc { |x| }.then { |l| [l.arity, l.parameters.length] }
=> [1, 1]
irb(main):002:0> proc { |*x| }.then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):003:0> proc { |**x| }.then { |l| [l.arity, l.parameters.length] }
=> [0, 1]
irb(main):004:0> proc { |x, *y| }.then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
irb(main):005:0> proc { |x, **y| }.then { |l| [l.arity, l.parameters.length] }
=> [1, 2]
irb(main):006:0> lambda { |x| }.then { |l| [l.arity, l.parameters.length] }
=> [1, 1]
irb(main):007:0> lambda { |*x| }.then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):008:0> lambda { |**x| }.then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):009:0> lambda { |x, *y| }.then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
irb(main):010:0> lambda { |x, **y| }.then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
irb(main):011:0> o = Object.new; def o.call(x); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [1, 1]
irb(main):012:0> o = Object.new; def o.call(*x); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):013:0> o = Object.new; def o.call(**x); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):014:0> o = Object.new; def o.call(x, *y); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
irb(main):015:0> o = Object.new; def o.call(x, **y); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
```

Reply:
Nice, thanks!